### PR TITLE
Use shared constant in pytorch decoder

### PR DESCRIPTION
**Pull request by mvafin**
### Details:
 - *Use shared constant in pytorch decoder*
 - *Remove ptr constant decoder*

### Tickets:
 - *104710*

### CODE REVIEW

File: src/bindings/python/src/openvino/frontend/pytorch/decoder.py [hunk 1]
```diff
@@ -36,18 +37,17 @@ def ivalue_to_constant(ivalue):
...
+        if ivalue.ndim == 0:
+            assert str(ivalue.dtype()) in pt_to_ov_type_map, "Type is not known"
+            ov_type = pt_to_ov_type_map[str(ivalue.dtype)]
+            ov_const = op.Constant(ov_type, Shape([]), [ivalue.item()])
+        else:
+            ivalue = ivalue.to(memory_format=torch.contiguous_format)
+            narr = ivalue.numpy(force=True)
+            if not narr.flags['C_CONTIGUOUS']:
+                narr = np.ascontiguousarray(narr)
+            ov_const = op.Constant(narr, shared_memory=True)
```
Comment (jiwaszki) [inline src/bindings/python/src/openvino/frontend/pytorch/decoder.py:50]:
```suggestion
            narr = ivalue.numpy(force=True)
            if not narr.flags['C_CONTIGUOUS']:
                narr = np.ascontiguousarray(narr)
            ov_const = op.Constant(narr, shared_memory=True)
```
Isn't later `ascontiguousarray` just enough in this case? Is Torch making it impossible to detach later without `.to()` call?

Same comment for later usage.

File: src/bindings/python/src/openvino/frontend/pytorch/decoder.py [hunk 2]
```diff
@@ -36,18 +37,17 @@ def ivalue_to_constant(ivalue):
...
-        except Exception:
-            # old variant that makes a slow data copying
-            warnings.warn("[ WARNING ] Constant wasn't able to convert from data_ptr.")
-            nvalues = ivalue.numpy()
-            ovtype = np_to_ov_type_map[str(nvalues.dtype)]
-            ovshape = PartialShape(nvalues.shape)
-            ov_const = op.Constant(ovtype, ovshape.get_shape(), nvalues.flatten().tolist())
+    if isinstance(ivalue, torch.Tensor):
+        if ivalue.ndim == 0:
+            assert str(ivalue.dtype()) in pt_to_ov_type_map, "Type is not known"
```
Comment (jiwaszki) [inline src/bindings/python/src/openvino/frontend/pytorch/decoder.py:42]:
```
>>> x = torch.Tensor(1)
>>> x.dtype
torch.float32
>>> x = torch.tensor(1)
>>> x.dtype
torch.int64
>>> x = torch.tensor(1.0)
>>> x.dtype
torch.float32
>>> x = torch.tensor(np.int32(1))
>>> x.dtype
torch.int32
```
Can this be better to refactor `pt_to_ov_type_map` to work with just `dtype` and fallback to "convert int32" in specific cases?

File: src/bindings/python/src/openvino/frontend/pytorch/decoder.py (hunk 1 above)
Comment (mvafin) [reply src/bindings/python/src/openvino/frontend/pytorch/decoder.py:50]: `.to(memory_format=torch.contiguous_format)` has nothing to do with `C_CONTIGUOUS` tensors. This is to work with models in `CHANNELS_LAST` memory format: https://pytorch.org/tutorials/intermediate/memory_format_tutorial.html

Comment (jiwaszki) [reply src/bindings/python/src/openvino/frontend/pytorch/decoder.py:50]: Got it! Thanks!

**Review by jiwaszki — APPROVED** [review]

File: src/bindings/python/src/openvino/frontend/pytorch/decoder.py (hunk 2 above)
Comment (eaidova) [inline src/bindings/python/src/openvino/frontend/pytorch/decoder.py:42]: I think it may be useful to provide which type comes in the assert message. Maybe can we add it?

**Review by eaidova — APPROVED** [review]

File: src/bindings/python/src/openvino/frontend/pytorch/decoder.py (hunk 2 above)
Comment (mvafin) [inline src/bindings/python/src/openvino/frontend/pytorch/decoder.py:42]:
```suggestion
            assert str(ivalue.dtype()) in pt_to_ov_type_map, f"Type is not known{ivalue.dtype()}"
```

File: src/bindings/python/src/openvino/frontend/pytorch/decoder.py [hunk 3]
```diff
@@ -265,29 +260,20 @@ def as_string(self):
     def _as_constant_tensor(pt_value: torch.Value):
         ivalue = pt_value.toIValue()
         if pt_value.isCompleteTensor():
-            try:
-                ivalue = ivalue.to(memory_format=torch.contiguous_format).detach().cpu()
-            except Exception:
-                warnings.warn("[ WARNING ] Tensor couldn't detach")
-            if str(pt_value.type().dtype()) in pt_to_ov_type_map:
+            if ivalue.ndim == 0:
+                assert str(ivalue.dtype) in pt_to_ov_type_map, "Type is not known"
```
Comment (mvafin) [inline src/bindings/python/src/openvino/frontend/pytorch/decoder.py:264]:
```suggestion
                assert str(ivalue.dtype) in pt_to_ov_type_map, f"Type is not known{ivalue.dtype}"
```

File: src/bindings/python/src/openvino/frontend/pytorch/decoder.py [hunk 4]
```diff
@@ -265,29 +260,20 @@ def as_string(self):
     def _as_constant_tensor(pt_value: torch.Value):
         ivalue = pt_value.toIValue()
         if pt_value.isCompleteTensor():
-            try:
-                ivalue = ivalue.to(memory_format=torch.contiguous_format).detach().cpu()
-            except Exception:
-                warnings.warn("[ WARNING ] Tensor couldn't detach")
-            if str(pt_value.type().dtype()) in pt_to_ov_type_map:
+            if ivalue.ndim == 0:
+                assert str(ivalue.dtype) in pt_to_ov_type_map, f"Type is not known{ivalue.dtype}"
```
Comment (eaidova) [inline src/bindings/python/src/openvino/frontend/pytorch/decoder.py:264]:
```suggestion
                assert str(ivalue.dtype) in pt_to_ov_type_map, f"Type is not known {ivalue.dtype}"
```

File: src/bindings/python/src/openvino/frontend/pytorch/decoder.py [hunk 5]
```diff
@@ -36,18 +37,17 @@ def ivalue_to_constant(ivalue):
...
-        except Exception:
-            # old variant that makes a slow data copying
-            warnings.warn("[ WARNING ] Constant wasn't able to convert from data_ptr.")
-            nvalues = ivalue.numpy()
-            ovtype = np_to_ov_type_map[str(nvalues.dtype)]
-            ovshape = PartialShape(nvalues.shape)
-            ov_const = op.Constant(ovtype, ovshape.get_shape(), nvalues.flatten().tolist())
+    if isinstance(ivalue, torch.Tensor):
+        if ivalue.ndim == 0:
+            assert str(ivalue.dtype()) in pt_to_ov_type_map, f"Type is not known{ivalue.dtype()}"
```
Comment (mvafin) [inline src/bindings/python/src/openvino/frontend/pytorch/decoder.py:42]:
```suggestion
            assert str(ivalue.dtype()) in pt_to_ov_type_map, f"Type is not known {ivalue.dtype()}"
```
